tests/rptest/tests/end_to_end.py

```diff
@@ -225,52 +237,112 @@ def run_validation(self,
             self.producer.stop()
             self.run_consumer_validation(
                 consumer_timeout_sec=consumer_timeout_sec,
-                enable_idempotence=enable_idempotence)
+                enable_idempotence=enable_idempotence,
+                enable_compaction=enable_compaction)
         except BaseException:
             self._collect_all_logs()
             raise
 
     def run_consumer_validation(self,
                                 consumer_timeout_sec=30,
-                                enable_idempotence=False) -> None:
+                                enable_idempotence=False,
+                                enable_compaction=False) -> None:
         try:
             self.await_consumed_offsets(self.producer.last_acked_offsets,
                                         consumer_timeout_sec)
             self.consumer.stop()
 
-            self.validate(enable_idempotence)
+            self.validate(enable_idempotence, enable_compaction)
         except BaseException:
             self._collect_all_logs()
             raise
 
-    def validate(self, enable_idempotence):
+    def validate_compacted(self):
+
+        consumer_state = {}
+
+        acked_producer_state = {}
+        not_acked_producer_state = defaultdict(set)
+
+        for k, v in self.producer.acked:
+            acked_producer_state[k] = v
+
+        # some of the not acked records may have been received and persisted,
+        # we need to store all of them and check if consumer result is one of them
+        for k, v in self.producer.not_acked:
+            not_acked_producer_state[k].add(v)
+
+        for k, v in self.records_consumed:
+            consumer_state[k] = v
+
+        msg = ""
+        success = True
+        errors = []
+
+        for consumed_key, consumed_value in consumer_state.items():
+            # invalid key consumed
+            if consumed_key not in acked_producer_state and consumed_key not in not_acked_producer_state:
+                return False, f"key {consumed_key} was consumed but it is missing in produced state"
+
+            # success case, simply continue
+            if acked_producer_state[consumed_key] == consumed_value:
+                continue
+
+            # we must check not acked state as it might have been caused
+            # by request timeout and a message might still have been consumed by consumer
+            self.logger.warn(
+                f"Checking not acked produced messages for key: {k}, previous acked value: {acked_producer_state[consumed_key]}, consumed value: {v}"
+            )
```

Comment:
nit: I imagine this is helpful for debugging, but it's possible this is totally fine and is just noise for our runner logs. Maybe bump it down to debug or info? Or move it below the `continue`, where it's actually problematic?

Reply:
done